### PR TITLE
Finitecell

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**
Multiple efficiency improvements:
- removes cone elements
- removes midpoints for ndims >= 2
- adds edge/vertex relations

### CODE REVIEW

File: nutils/element.py [hunk 1]
```diff
@@ -241,13 +238,15 @@ def slice(self, levelfunc, ndivisions):
...
             midpoint = v0 + xi * (v1-v0)
         else:
-            nbins = 2**ndivisions
-            clevel = levelfunc(self.centroid[_])[0]
-            select = clevel*levels <= 0 if clevel != 0 else levels != 0
-            levels = levels[select]
-            vertices = self.vertices[select]
-            xi = numpy.round(levels/(levels-clevel) * nbins)
-            midpoint = numpy.mean(vertices + (self.centroid-vertices)*(xi/nbins)[:, _], axis=0)
+            for trans, edge, emap, newedge in zip(self.edge_transforms, self.edge_refs, self.edge_vertices, refs):
```
Comment (joostvanzwieten) [inline nutils/element.py:209]: I understand that you select a new vertex as midpoint, or if there is none a vertex of the base that is a vertex of exactly one of all trimmed edge simplices. But why?

**Review by joostvanzwieten — CHANGES_REQUESTED** [review]

File: nutils/element.py [hunk 2]
```diff
@@ -191,6 +191,13 @@ def slice(self, levelfunc, ndivisions):
             return self
         if numpy.less_equal(levels, 0).all():
             return self.empty
+        assert self.ndims >= 1
+
+        refs = [edgeref.slice(lambda vertices: levelfunc(edgetrans.apply(vertices)), ndivisions) for edgetrans, edgeref in self.edges]
+        if sum(ref != baseref for ref, baseref in zip(refs, self.edge_refs)) < self.ndims:
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py:221]: done 064418a6

File: nutils/element.py [hunk 3]
```diff
@@ -214,19 +214,17 @@ def slice(self, levelfunc, ndivisions):
         if sum(bool(ref) for ref in refs) < self.ndims:
             return self.empty
 
-        nbins = 2**ndivisions
-
         if self.ndims == 1:
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py:226]: The old message referred to the input of the slice operation, which at that point still has two vertices, but the new message should be a lot clearer. b7ec9723

File: nutils/element.py [hunk 4]
```diff
@@ -214,19 +214,17 @@ def slice(self, levelfunc, ndivisions):
...
-            xi = numpy.round(l0/(l0-l1) * nbins)
-            if xi in (0, nbins):
-                return self.empty if xi == 0 and l1 < 0 or xi == nbins and l0 < 0 else self
-            v0, v1 = self.vertices
-            midpoint = v0 + (xi/nbins) * (v1-v0)
-
+            iedge = [i for (i,), edge in zip(self.edge_vertices, self.edge_refs) if edge]
+            l0, l1 = levels[iedge]
+            nbins = 2**ndivisions
+            xi = round(nbins * l0 / (l0-l1)) / nbins
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py]: Done (reverted old code)

File: nutils/element.py [hunk 5]
```diff
@@ -35,6 +35,21 @@ def __init__(self, ndims: int):
     def nverts(self):
         return len(self.vertices)
 
+    @property
+    def vertices(self):
+        raise NotImplementedError(self)
+
+    @property
+    def edge_vertices(self):
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py:74]: Done, with some tweaks. Is the `Sequence[Sequence[int]]` part commonly found in a doc string? Hadn't seen that before, I left it out for now (same below).

File: nutils/element.py [hunk 6]
```diff
@@ -315,6 +319,10 @@ def vertices(self):
     def edge_vertices(self):
         return self.baseref.edge_vertices
 
+    @property
+    def simplices(self):
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py:342]: Done with tweaks.

File: tests/test_element.py [hunk 7]
```diff
@@ -77,6 +77,14 @@ def test_edge_vertices(self):
...
             self.assertAllEqual(self.ref.vertices[everts], etrans.apply(eref.vertices))
 
+    @parametrize.enable_if(lambda ref, **kwargs: not isinstance(ref, element.WithChildrenReference) and ref.ndims >= 1)
+    def test_simplices(self):
+        volume = 0
+        for simplex in self.ref.vertices[self.ref.simplices]:
+            simplex_volume = numpy.linalg.det(simplex[1:] - simplex[0]) / math.factorial(self.ref.ndims)
+            self.assertGreater(simplex_volume, 0)
+            volume += simplex_volume
+        self.assertAlmostEqual(volume, self.ref.volume)
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply tests/test_element.py:86]: I added a centroid test which amounts to precisely that.

File: nutils/element.py [hunk 8]
```diff
@@ -43,6 +43,11 @@ def vertices(self):
     def simplices(self):
         raise NotImplementedError(self)
 
+    @property
+    def simplex_transforms(self):
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py:53]: Done with tweaks.

File: nutils/element.py [hunk 9]
```diff
@@ -1033,50 +1033,62 @@ def __init__(self, baseref, edge_refs: tuple, midpoint: types.arraydata):
...
+                edge_vertices.append(emap)
+
+            orientation = []
+            seen = {}
+            for edge1, newemap1, etrans1, newedge1 in zip(baseref.edge_refs, edge_vertices, baseref.edge_transforms, edge_refs):
+                newedge1_edge = zip(newedge1.edge_vertices, newedge1.edge_transforms, newedge1.edge_refs)
+                trimmed = []
+                for edge2, (newemap2, etrans2, newedge2) in zip(edge1.edge_refs, newedge1_edge):
+                    if edge2: # existing non-empty edge
+                        key = tuple(newemap1[newemap2[edge2.simplices]].flat) # we look at simplices to disregard unused vertices
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py]: Having worked through the entire logic once again, this time leaving extensive comments in place, I think sorting is not required. 03ba63c4

File: nutils/element.py [hunk 10]
```diff
@@ -1092,40 +1092,38 @@ def __init__(self, baseref, edge_refs: tuple, midpoint: types.arraydata):
 
         super().__init__(baseref.ndims)
 
+    def _with_edges(self, edge_refs):
```
[the remark this answers is not in the document]
Comment (gertjanvanzwieten) [reply nutils/element.py:1044]: What about the third branch though, which requires `self._midpoint`?

File: nutils/element.py (hunk 1 above)
Comment (gertjanvanzwieten) [reply nutils/element.py:251]: New comments hold the answer 2ec17eff

File: nutils/element.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (joostvanzwieten) [reply nutils/element.py:1044]: I missed that.

**Pull request merged**